In [2]:

import ipynbname
from jupyter_lab_notebook_toc_utils import generate_toc, display_toc
toc = generate_toc(ipynbname.name() + ".ipynb", add_numbering=True)
display_toc(toc)

**Table of Contents**<br/>
&nbsp;&nbsp;&nbsp;&nbsp; [**1.** **Overview**](#Overview)<br/>
&nbsp;&nbsp;&nbsp;&nbsp; [**2.** **Major Topics**](#Major-Topics)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**2.1.** Alignment & Alignment Score](#Alignment-&-Alignment-Score)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**2.2.** Annotations](#Annotations)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**2.3.** Sigmoid Function](#Sigmoid-Function)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**2.4.** Softmax Function](#Softmax-Function)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**2.5.** Attention Layers](#Attention-Layers)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**2.6.** Attention Vectors](#Attention-Vectors)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**2.6.1.** Connection With Context Vectors](#Connection-With-Context-Vectors)<br/>
&nbsp;&nbsp;&nbsp;&nbsp; [**3.** **How Attention Works (Conceptually)**](#How-Attention-Works-%28Conceptually%29)<br/>
&nbsp;&nbsp;&nbsp;&nbsp; [**4.** **Types Of Attention**](#Types-Of-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.1.** Implicit vs Explicit](#Implicit-vs-Explicit)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.2.** Hard Vs. Soft](#Hard-Vs.-Soft)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.3.** Local vs. Global](#Local-vs.-Global)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.4.** Self vs. Cross Attention (inter vs. intra)](#Self-vs.-Cross-Attention-%28inter-vs.-intra%29)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.5.** Additive (Bahdanau) Attention](#Additive-%28Bahdanau%29-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.5.1.** TODO: Fix notation and explain how the alignment function connect to Softmax. See Self-attention notebook](#TODO:-Fix-notation-and-explain-how-the-alignment-function-connect-to-Softmax.-See-Self-attention-notebook)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.6.** Dot-Product Attention](#Dot-Product-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.7.** Scaled Dot-Product Attention](#Scaled-Dot-Product-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.8.** QKV Attention](#QKV-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.8.1.** Etymology](#Etymology)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.8.2.** Implementation](#Implementation)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.8.3.** Visualization](#Visualization)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.8.4.** Intuition](#Intuition)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.8.5.** Connection With Prior Works](#Connection-With-Prior-Works)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.9.** Multi-Head attention](#Multi-Head-attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.9.1.** Etymology](#Etymology)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**4.10.** Multihead QKV Attention](#Multihead-QKV-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp; [**5.** **Motivation**](#Motivation)<br/>
&nbsp;&nbsp;&nbsp;&nbsp; [**6.** **History**](#History)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.1.** Graves et. al. (2014) - Birth of Attention](#Graves-et.-al.-%282014%29---Birth-of-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.2.** Kalchbrenner et. al., (2014) - Variable Input Length Using CNNs](#Kalchbrenner-et.-al.,-%282014%29---Variable-Input-Length-Using-CNNs)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.3.** Mnih et. al. (2014) - Attention Based RNNs for Image Processing](#Mnih-et.-al.-%282014%29---Attention-Based-RNNs-for-Image-Processing)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.4.** Cho et al. (2014) - Performance Deteriorates With Sequence Length](#Cho-et-al.-%282014%29---Performance-Deteriorates-With-Sequence-Length)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.5.** Cho, Bengio, and Bahdanau (2014) - Decoder Attention For RNNs](#Cho,-Bengio,-and-Bahdanau-%282014%29---Decoder-Attention-For-RNNs)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.6.** Bahdanau et. al. (2015) - Global Hard And Soft Alignment & Additive Attention](#Bahdanau-et.-al.-%282015%29---Global-Hard-And-Soft-Alignment-&-Additive-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.7.** Xu et. al. (2015) - Applying Soft/Hard Attention To Image Processsing](#Xu-et.-al.-%282015%29---Applying-Soft/Hard-Attention-To-Image-Processsing)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.8.** Luong, et al. (2015) - Global vs. Local Attention](#Luong,-et-al.-%282015%29---Global-vs.-Local-Attention)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.9.** Britz et. al (1017) - Global Soft Attention Catches On](#Britz-et.-al-%281017%29---Global-Soft-Attention-Catches-On)<br/>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [**6.10.** Vaswani et al (2017) - Transformers and Attention](#Vaswani-et-al-%282017%29---Transformers-and-Attention)

# Overview

In machine learning, most commonly deep learning, attention is the anthropomorphization of the concept of human attention. In other words, it gives the machine learning algorithms an ability to selectively focus on certain pieces of information while ignoring other pieces of information.

Attention generally involves assigning weights to information which indicate importance; zero being not important and one being the most important.

# Major Topics

## Alignment & Alignment Score
While doing research for this notebook, the term alignment came up a number of times. In the context of machine translation and related NLP tasks, the term referse to the conceptual mapping, or alignment of terms from an input sequence to an output sequence.

Conceptually, an alignment score gives a numerical value which measures the alignment of two terms or sequences. Practically speaking, this is a measure of about how relevant, or strongly correlated, the terms are to eachother and thus how much attention we should pay to them. A higher value indicates that these terms will likely affect eachother during alignment an then should be considered by the attention mechanism. An allignment score of zero means that the terms are unrelated and not relevant to the alignment process thus we should not pay attention to them.

The alignment score may refer to the weights applied to the thing being attended, or they may be a component of an attention score as see in Cho, Bengio, and Bahdanau (2014).

For more information see this article and the Attention notebook.

For more infomration see the [Extra Topics notebook](ExtraTopics.ipynb#Alignment).

## Annotations

The term annotations refers to a component of an attention mechanism; specifically the concatenation of hidden states within a network that is being attended.

For more infomration see the [Extra Topics notebook](ExtraTopics.ipynb).

## Sigmoid Function

The sigmoid function plots a continuous s-shaped curve whos range is bound between 0 and 1.

The properties of the sigmoid function are useful for classification problems. 

For more information see the [sigmoid function notebook](../../../../Mathematics/Functions/Sigmoid%20Functions.ipynb).

## Softmax Function

The Softmax function is a multidimensional vector function which transforms a vector such that each dimension is given a number between 0 and 1. This characteristic allows the function to be used in multinominal classification problems (i.e. classification problems involving more than two mutually exclusive classes).

For more information see the [softmax function notebook](../../../../Mathematics/Functions/Softmax%20Function.ipynb).

## Attention Layers

In a number of papers and blogs [(example)](https://dmol.pub/dl/attention.html#attention-layers) I see authors referencing "the attention layer" or pointing out that attention can be added to a neural network via a layer or collection of layers. 

Initially I found this confusing until I studdied an annotated [reference](https://analyticsindiamag.com/a-beginners-guide-to-using-attention-layer-in-neural-networks/) to a diagram to [Cho et. al. (2024)](https://arxiv.org/abs/1409.0473)

<center><img src='./images/annotated_attention_network_cho_2014.png' style='width:600px'></center>

Here, in red, we can see a "slice" of a feed forward network. In the dotted red box, we see all the the connections from a hidden layer ($h_t$) to a single node in another level ($y_t$). If we were to draw the connection for the other nodes ($y_t$) in the lower level we would see a normal diagram for a fully connected feed-forward network layer as shown below:

<center><img src="images/dense_connection_diagram.png" style='width:300px'></center>

**Note**: In this diagram the grey nodes are illustrating the same "slice" as depicted in the first diagram without needed to omit the connection lines. I find this to be a more clear illustration for the novice reader.

While there are many implimentations of attention, its possible that we may use more than a single layer to perform the attention mechanism.

## Attention Vectors

Generally speaking, the term attention vector typically refers to a vector quantifies the importance, or the amount of attention which should be paid to the individual terms within the input sequence while an attentive model is performing various tasks (such as translation, summarization, etc.).

As many attention mechanisms are implimented as neural networks, the attention vector will refer to the output of the output layer.

### Connection With Context Vectors

Context vectors, or word embeddings, are mathematical representations for the meaning of a word. The dimensions of these vectors each corespond to a different dimension of meaning. A higher score in a given dimension indicating that a word has a strong relationship or association with the connotations of the dimension. In some implimentations, the dimensions of the vector map to words, either from a unique set of the words that appear in a corpus.

As the conceptualization of these vectors becomes more abstract, the meaning of a given dimension becomes less straight forward. In many approaches to embeddings, these dimentions are intentionally denoted as not having explicit a priori meaning that can be intepreted clearly by a human.

In modern implimentations, context is thought of as a fucntion of the co-oocurance of terms within the input sequence. As such, unique context vectors are produced for each unique input sequence. This transition marks a convergence with various implimentations of attention and self-attention.

As noted by [Britz et. al. (2017)](https://arxiv.org/abs/1703.03906), the terms context vector and attention vector are used synonomously and is calculated as a weighted average 
of the source state (i.e. the input sequence hidden states).

# How Attention Works (Conceptually)

The purpose of an attention mechanism is to indicate how important a given piece of information is to a task being performed by a model.

Common exmamples of NLP seq2seq tasks that an attentive model might perform include translation, summarization, or Q&A. In this context, the attention mechanism will rank the interdependence or correlation between the tokens in the input sequence and the output sequence.

Bahdanau et. al. (2014) provide a helpful visualization of attention scores derived for a translation model. We can see how the attention scores allow the model to dynamically determine the soft-alignment of the input and output sequences. Specifically, they show several instances where the attention model is able to change the order of words from their exact literal translation to a gramatically correct translation.

<center><img src="./images/attention_swapping_word_order.png" style="width:50%"></center>

> Figure 3: Four sample alignments found by RNNsearch-50. The x-axis and y-axis of each plot
correspond to the words in the source sentence (English) and the generated translation (French),
respectively. Each pixel shows the weight $\alpha_{ij}$ of the annotation of the j-th source word for the i-th
target word (see Eq. (6)), in grayscale (0: black, 1: white). (a) an arbitrary sentence. (b–d) three
randomly selected samples among the sentences without any unknown words and of length between
10 and 20 words from the test set.
 


# Types Of Attention

There are many types of attention mechanisms which have evolved over the last ten years. Many mechanisms are classified by their architecture or some other physical characteristic while a few are classified by how they are applied or what they are attending regardless of their architecture.

This [article](https://theaisummer.com/attention/#the-limitations-of-rnns) was a helpful in providing a high level overview.

## Implicit vs Explicit

I believe these terms, in the context of data science, are analogs of their neuro-biological counterparts dealing with actual physical human attention mechanisms (for example [this article](https://www.frontiersin.org/articles/10.3389/fpsyg.2015.01861/full). From what I understand, one way of thinking about attention is to classify the attention mechanism as either implicit or explicit.

Explicit attention mechanisms are voluntarily focused on goal relevant stimuli from the environment. For example, if we are looking for pictures of vehicles trying to classify them as car or truck, we will be focusing on the groups of pixels in the image most related to the classification problem; more likely, we will be focusing on the gradient of the loss function. Thinking about translation on the other hand, explicit attention would focus on the words most relvant to the proper translation.

Implicit attention mechanisms are involunarily focused on stimuli who's inherant properties inadvertantly manipulate the attention mechanism. In the human analogy, the images may invoke an emotional response that causes us to focus on features of the stimuli which are irrelevant to the stated goal of classification. In the literal case of deep neural networks, it's possible that the physical structure of the network impact the attention mechanism without a direct causal relationship being defined a priori with respect to the stated goal. Thinking about translation, it may be that a particular passage or set of words comes into focus because of some conincidental relationship with the model's current state or architecture.

Putting it simply, With explicit attention, the physical system is deigned to adjust the attention mechanism based on the model performance during training. With implicit attention, the attention mechanism is not explicitly alterered as part of the training process; instead any changes to the attention mechanism are an unintentional byproduct.

Here is a [link](https://www.linkedin.com/pulse/attention-mechanism-part-1-english-version-hay-hoffman/) to an article worth reading.

## Hard Vs. Soft

Generally speaking, at the core of every attention mechanism is a function which assigns attention scores. These functions, and thus the attention mechanisms, are classified as being either soft or hard. Soft attention functions ar characterized by their continuous, smooth, differentiable nature. Hard attention functions are discrete, non-smooth, and non-differentiable.

https://theaisummer.com/attention/#types-of-attention-hard-vs-soft

The characteristic of being differentiable is important with respect to neural network archiectecure because back propogation does not work if a function is non-differentiable.

https://stats.stackexchange.com/questions/386535/why-cant-we-use-back-propagation-in-hard-attention-but-we-can-use-it-in-relu#

## Local vs. Global

Global attention describes attention mechanisms that consider all input stimuli (e.g. pixels or words) while local attention considers a curated set of words as instructed by the training process.

> Luong, et al., 2015 proposed the “global” and “local” attention. The global attention is similar to the soft attention, while the local one is an interesting blend between hard and soft, an improvement over the hard attention to make it differentiable: the model first predicts a single aligned position for the current target word and a window centered around the source position is then used to compute a context vector.
>
> [source](https://lilianweng.github.io/posts/2018-06-24-attention/)

## Self vs. Cross Attention (inter vs. intra)

As noted in the [Self Attention notebook](Self-Attention.ipynb), Self-attention
(also known as intra-attention) is a type of attention that
quantifies the interdependence within the elements of a single input. At the same time, cross-attention (also known
as inter-attention) finds the interdependence across two or
more inputs.

## Additive (Bahdanau) Attention

Additive attention refers to an attention mechanism which calculates a context vector $c$ as the aggregation of $\alpha$ weighted annotations $h_i$, or hidden states $\vec h_i$ more generally. This mechanism is recognized by it's mathematical formula

$$ c_i = \sum \alpha_i h_i $$

which described in Bahdanau et. al. (2015) and Cho, Bengio, and Bahdanau (2014).

In these papers we see this mechanism being applied to RNNs as a global attention mechanism.

Being that Bahdanau is the common factor, we see some refer to this as Bahdanau attention.

### TODO: Fix notation and explain how the alignment function connect to Softmax. See Self-attention notebook

## Dot-Product Attention

With dot product attention, we are summing the weighted annotations. As [Luong, et al. (2015)](https://arxiv.org/abs/1508.0402) points out, with dot-product attention we are multiplying (taking the dot product of) the hidden states rather than weighting and adding them.

$$score(h_t, \bar h_s) = h_t^T \bar h_s $$

Where $h_t$ refers to the target hidden state (i.e. the encoder's hidden state) and $\bar h_s$ refers to the source hidden state (i.e. the decoder's hidden state).

Luong et. al (2015) also point to a "general" attention which is a weighted version of Dot-Product attention. In the special case, where the weight is a constant scaling term we have Scaled Dot-Product attention.

## Scaled Dot-Product Attention

**Note**: I have seen claims referring to this type of attention as being self-attention. These claims are false. The author is likely conflating the concept of self-attention with the implimentation of a self-attention that uses a scaled dot-product calculation as popularized by Vaswani et. al. (2007).

As discussed in this [source](https://www.educative.io/answers/what-is-the-intuition-behind-the-dot-product-attention): 

> The scaled dot product attention mechanism is a subset of the dot product attention mechanism. It uses a dot product operator with scaling to maximize training data and stability by ensuring the dataset is optimized for the model during training.


We can formulate this attention score as follows: 

$$score(h_t, \bar h_s) = \frac{h_t^T \bar h_s}{c} $$

A common paper which impliments a form of scaled dot-produt attention is [Vaswani et. al. (2007)](https://arxiv.org/abs/1706.03762) and is often cited as the definition of scaled dot-product attention (which it's not). However this implimentation has three components being attended rather than the typical two. 

According to [Vaswani et. al. (2007)](https://arxiv.org/abs/1706.03762), the scaled dot-product attention is not a huge departure from the common attention mechanisms of the time. They inform us that, as the name suggests, this is just a scaled version of dot-product attention:

> The two most commonly used attention functions are additive attention [2], and dot-product (multiplicative) attention. Dot-product attention is identical to our algorithm, except for the scaling factor
>
> ...
>
> We call our particular attention "Scaled Dot-Product Attention" (Figure 2)

They give the following formula for their attention mechanism:

> $$ \textrm{Attention}(Q, K, V) = \textrm{softamax} \left(\frac{QK_T}{\sqrt{d_k}} \right)V$$

We can see at it's core, it is performing dot-product attention, but it also have a number of significant differentiators.

[Vaswani et. al. (2007)](https://arxiv.org/abs/1706.03762) also provide the following diagram to illustrate the high level process by which this computation is performed:

<center><img src="images\scaled_dot_product_attention.png"></center>

Here, the inputs are matrices Q,K,V and the outputs are an attention score matrix. The Matmul block refers to the process of multiplying matrices, scale refers to the division by the scaling factor, etc.

For more information about how Vaswani et. al. (2017) use this mechanism, see the [Transformer notebook](Transformers.ipynb).

Additionally Vaswani et. al. (2017) refer to their attention mechanism as an "attention layer" implying that what we are looking at fits into the activation function of the neuron in the neural network.

## QKV Attention

### Etymology

While they do not use the term, [Vaswani et. al. (2007)](https://arxiv.org/abs/1706.03762) are responsible for the publication if this attention mechanism. 

**Note**: Many articles will cite this as THE definition of self attention. This is incorrect. Self attention existed before and after this paper was authored. This mechanism is simply one very popular implimentation of self attention.

**Note**: It's also important to understand that this mechanism can be used for both intra and inter attention as was the case in Vaswani et. al. (2007).

As noted [here](https://stats.stackexchange.com/questions/421935/what-exactly-are-keys-queries-and-values-in-attention-mechanisms), 

> For unsupervised language model training like GPT, Q,K,V
 are usually from the same source, so such operation is also called self-attention.

### Implementation

Like many attention mechanisms, this one is implimented as a neural network. It is common for these types of NN-based mechanisms to be classified by the nature of the activation function in their output layer. As the output layer is what provides the output (i.e. the attention vector) it is largely the activation function of this layer which defines the calculates for the attention vector.

As such, this attention mechanism gets its name due to the fact that it's activation function relies on three logical constructs. These are referred to as Querys, Keys, and Values and are denoted as Q, K, and V, resepctively.

The authors give the following definition for the activation function:

$$ Attention(Q, K, V) =  softmax \big( \frac{QK^T}{\sqrt{d_k}} \big)V $$

and provide the following functional diagram to desribe and sequence the calculations involved in the computation of the activation function:



<center><img src="images\scaled_dot_product_attention.png"></center>

While not explicitly stated in their paper, I was able to deduce that the $Q$, $K$, and $V$ matrices are produced by multiplying the input embeddings $X$ by a coresponding weight matrices, $W^Q$, $W^K$, and $W^V$ which corespond to the Query, Key, and Value respectively.

### Visualization

The attention mechanism is designed to work with embeddings not raw inputs. As such we must first calculate the word embeddings for the given input. For example, the french phrase "I am a student" or "je suis etudient" would be broken down into embedding vectors.

<center><img src="./images/transformer_word_embeddings_example2.png" style="width:75%"></center>

<div style="text-align:right"><a href="https://jalammar.github.io/illustrated-transformer">[img source]</a><div>

These embeddings would are stored and manipulated in a matrix $X$ for mathematical convenience. The deimensionality of this matrix are (token count x embedding length).

Below we can see an example of a matrix which contains the embedding vectors for each word in the given input sequence "Thinking Machines":
    
<center><img src="./images/transformer_word_embeddings_example.png" style="width:30%"></center>

    
<div style="text-align:right"><a href="https://jalammar.github.io/illustrated-transformer">[img source]</a><div>

Next is to calculate three intermediary matrices $Q$, $K$, and $V$ which are referred to as they query, key, and value vectors respectively. 

As discussed [here](https://jalammar.github.io/illustrated-transformer/) and [here](https://stackoverflow.com/questions/68266490/dimension-of-query-and-key-tensor-in-multiheadattention), these matrices are produced by applying linear transformations ($W_K$,$W_Q$, and $W_V$) to the input embeddings $X$:

$$ X W_Q = Q $$
$$ X W_K = K $$
$$ X W_V = V $$

The letter $W$ was specifically selected to denote or imply that these values are derived through the co-training process of the attention mechanism. Recall that the attention mechanism, a neural network, is co-trained with the encoder and decoder. As such, it's hidden layer consists of weights which are adjusted through the back-propogation.

This is a very important realization as it intuitively allows us to understand how and why the Qeury, Key, Value actually work together: The loss function and training process calibrate them to.

**Note**: The dimensionality of the linear transformation matrices and thus the resulting query, key, and value matrices is flexible (i.e. The number of columns can change). Some architectures elect specific dimensions so that the calculations take on certain characteristics or properties for ease of use or speed, etc.

That being said, concuptually, I think it makes sense to think of them as having the same dimensionality of $X$ (token count x embedding length). Thus every embedding $x_i$ has a coresponding query $q_i$, key $k_i$, and value $v_i$ respectively.

<center><img src="./images/transformer_query_key_and_value_example2.png" style="width:60%"></center>

We can then plug these matrices into the definition of attention:

$$ Attention(Q, K, V) =  softmax \big( \frac{QK^T}{\sqrt{d_k}} \big)V $$

Visually, we can see the dimensionality:

<center><img src="./images/transformer_attention_example.png" style="width:50%"></center>

Focusing on the numerator we can visualize the dot product vector multiplication that is being utlized:

<center><img src="./images/transformer_attention_example2.png" style="width:50%"></center>


Becasue the query and key matrices have the same dimensions, the result of the calculation is a square matrix. Here each element represents the dot-product attention score between the query and the key.

Continuing the calculation gives us:

<center><img src="./images/transformer_attention_example3.png" style="width:50%"></center>

<center><img src="./images/transformer_attention_example4.png" style="width:20%"></center>


The softmax function plays an important role in this computation. It ensures that the dimensions of the resulting calculation each sum to one. As such, each element in the result matrix is interpreted as a weight or probability; the probability that the scaled dot-product of the query and key coresponds or relates to the given value.

As noted earlier, through training the values of the coresponding weight matrices are optimized so that the probability associated with a given query and key is the highest for the most relevant value embedding.

### Intuition

So how do these three constructs work? 

To answer this question we must first absrtact this attention function as a sort of expectation. We can consider the softmax function to be a sort of probability function and we can train the weight matrices $W^Q$ and $W^K$ such that they maximize the probability returned by the softmax function when the Qeury and Key are relevant to the given Value. If we multiply a Value by it's probabilty of occrring we get a sort of expectation value.

$$ =  softmax \big( W \big)V $$

$$ = pV $$


In the case of inter attention, or self attention, the Q, K, and V are all derived from the same input sequence. Thus the probability vector represents perceived correlation or relevance between the terms in the input sequence. And the product represents a projection of the Value into the continuous Value space; Values being geometrically closer being more simillar.

In the case of intra attention, or cross attention, the Q and K are derived from the input sequence while the value is relating to the output sequence. Thus the decoder is able to calculate probabilistic relationships between the input and output and also to train its probabilistic model. In this case, the probability would again represent the perceived correlation or relevance and the product would again represent a comparable projection.

Given this probabilistic model, and continuous embedddings, [some](https://stats.stackexchange.com/questions/421935/what-exactly-are-keys-queries-and-values-in-attention-mechanisms) note that this attention mechanism can be thought of as being analogous to a retrieval system

> The key/value/query concept is analogous to retrieval systems. For example, when you search for videos on Youtube, the search engine will map your query (text in the search bar) against a set of keys (video title, description, etc.) associated with candidate videos in their database, then present you the best matched videos (values).



If the probablity weights are restricted to a one-hot vector (composed of zeros and only one one) we could conceive that this is actually preforming retrival as only a single value would be returned.

**Note**: Personally, I found this helpful at first, but a bit too simple the more I thought about it. My main issues was that with continuous embeddings, there is not a finite store of documents being mapped to keys or metadata, but a continuous space being mapped to words (in the case of translation). I couldnt understand how the projection of the Value would translate back into the input or output space; it's conceivable that no such word or phrase exists at the projected vecros geometric location within the Value space. But I reason that the mapping function would handle any of these issues by picking a nearest neighbor so to speak.

### Connection With Prior Works
So what differentiates this attention mechanism from prior works?

Generally speaking, there are several ways to think about conenctions with prior works, we have the mathematical structure of the activation function, we have the inputs to the function, and we have the usage of the mechanism.

For example, recall the prior work of [Bahdanau et. al. (2014)](https://arxiv.org/abs/1409.0473) had a similar mathematical structure. However we will see that this structure relied on a less abstract mechanism. While QKV attention relied on the hidden states directly, QKV relied on a more expressive abstraction applied raw data. 

<center><img src="images\attention_architecture_example.png"></center>


Here, the context vector $\bigoplus$ for the encoder-decoder is computed by multiplying the annotation $h$ (a concatenation of the forward and backward hidden states of the encoder) by a weight vector $\alpha$ and then summing the values for each position in the $T_x$ length input vector $x$. 

This is represented mathematically as:
$$ c_i = \sum_j^{T_x}{\alpha_{ij} h_j} $$ 

Where the weight $\alpha$ is computed as the softmax function being applied to an alignment model $a$. Thus, abstractly, we would have:

$$ = \sum_j^{T_x}{\textrm{softmax}(a_{i,j}) h_j}$$
$$ = \sum_j^{T_x}{p h_j}$$

Here we would again arrive at a sort of expectation for the hidden states (annotation $h$) given some sort of input (in this case the alignment between input and output so we are talking about cross attention).

## Multi-Head attention

### Etymology

The choice for the term "attention head" is not made explicitly clear. Additionally I was unable to definitively determine who coined the term. The term is made popular by Vaswani et. al. (2017) and as far as I can tell, they are the first to apply a multi-head design to an attention mechanism. 

I believe the term "head" is deriving it's meaning analogously from the field of electronics. In electronics, the part of a recording device (like a hard drive) that senses data encoded on a medium (like the hard drives pallet or magnetic tape) is referred to as a "read head". In the context of attention, the attention head is behaving similarely in that it is reading infomration (i.e. the input sequence) into the attention mechanism. As such, the head is performing an attention operation rather than a simple read operation which is why I belive it is called attention head rather than read head.

Taking this anolog a step further, the general concept of a multi-head architecture is comparatively old. A multi-head mechanism has multiple heads that can each independently perform the requested opetations. We see this being implimented in storage solutions which have multiple read/wrtie heads to maximize the input/output operations per second (IOPS). 

We can infir, and confirm through analysis of Vaswani et. al. (2017), that the multihead attention mechanism allows the transformer to apply multiple attention heads (i.e. calibrate multiple attention functions) in parallel for the same input sequence.

## Multihead QKV Attention

With single-head attention, for each input token in the $d_{model}$ length input sequence, a coresponding Query, Key, and Value are calculated. If we wanted to train multiple attention mechanisms on this same sequence we would need to serially employ the same mechanism and this would be inefficient.

With multi-head attention, these Queryies, Keys, and Values are linearly projected $h=8$ times creating multiple separate sets of queries, keys, and values. The authors denote the projections as $W^Q_i$, $W^K_i$, and $W^V_i$ respecively.

**Note**: These projections are different from the weights mentioned in the single-head attention formulae.

The authors provide the following functional diagram to visually describe the computation:

<center><img src="./images/multi_head_attention.png" style='width:300px'></center>

The multi-head attention mechanism is then able to apply the single-head attention mechanism to each of these linear projections, but in parallel. 

The "magic" in all this is that the projections, like the weight matrices in single-head attention are learned and calibrated by the training data. Additionally, because they are all different, they are each attenting to different parts of the input sequence.

The authors insist that this implimentation is more robust than the single-head conterpart:

> Multi-head attention allows the model to jointly attend to information from different representation
subspaces at different positions. With a single attention head, averaging inhibits this.

The authors give the following definition for an attention head:

> $$ \textrm{head}_i =  \textrm{Attention}(QW^Q_i , KW^K_i , V W^V_i ) $$

Then, at the end, these projections are then concatenated and again projected by $W^O$.

The following formula is presented for the multihead attention.

$$ \textrm{MultiHead}(Q,K,V)=\textrm{CONCAT}(\textrm{head}_1, \dots, \textrm{head}_h)W^O $$

# Motivation

The concept of attention was born from a long and iterative attempt to solve several related problems. The timeline is a bit messy and convoluted, but ultimately the invention was motivated by the realization that models representing things like language required an overwhelming amount of contextual information to be considered while only having a limited practical amount of resources to perform the considerations.

The attention mechanism provides the algorithmic way to still have access to an overwhelming amount of information but only consider the limited subset of information which is deemed to be relevant to the particular considersation.

Advancements in Topic Modeling, Encoder-Decoder architectures, and Machine Translation (specifically Neural Machine Translation), together set the stage for the discovery of the modern concept of attention and its implimentation.

# History

Arround the 2014 era, we started to see a switch from CNNs to RNNs before in 2017 seeing the birth of the Transformer. During this time frame researchers were struggling to get their unsupervised models to be able to recognize and model a richer contextual meaning based on a growing volume of data while still being computationally efficient to be of practical use.

One of the major problems that was overcome during this time period was the problem of a fixed length input buffer. Additionally, the problem of convolutions and recurrent connections was also overcome.

## Graves et. al. (2014) - Birth of Attention

In October 2014, Graves, Wayne, and Danihelka [published](https://arxiv.org/abs/1410.5401) *Neural Turing Machines*.

According to this [article](https://slds-lmu.github.io/seminar_nlp_ss20/attention-and-self-attention-for-nlp.html):

> Attention was in the beginning not directly developed for RNNs or even NLP. Attention was first introduced in Graves, Wayne, and Danihelka (2014) with a content-based attention mechanism ($score(s_t,h_i) = cosine[s_t, h_i]$) for Neural Turing Machines. Their application for NLP related tasks were later developed by Luong, Pham, and Manning (2015), Bahdanau, Cho, and Bengio (2014) and Xu et al. (2015).

## Kalchbrenner et. al., (2014) - Variable Input Length Using CNNs
In April of 2014, a paper titled *"A Convolutional Neural Network for Modelling Sentences
"* was [published](https://arxiv.org/abs/1404.2188).

In this paper, Kalchbrenner et. al. put forward a novel convolutional architecture dubbed the Dynamic Convolutional Neural Network (DCNN).

The notariaty of this archectecutre is that it can handle input sentences of varying length. The variable length inputs are addressed by sequentially processing one token at a time, or by specially designed convolution kernels Additionally the model is capable of capturing short and long-range relations.

> This approach can lead to significant problems when the input is truly of varying size with varying information content, such as in Section 10.7 in the transformation of text (Sutskever et al., 2014). In particular, for long sequences it becomes quite difficult to keep track of everything that has already been generated or even viewed by the network. Even explicit tracking heuristics such as proposed by Yang et al. (2016) only offer limited benefit.
>
> https://d2l.ai/chapter_attention-mechanisms-and-transformers/queries-keys-values.html

Additionally, this sequential processeing leads to a performance hit compared to modern approaches which are able to compute in parallel.

## Mnih et. al. (2014) - Attention Based RNNs for Image Processing

In June 2014, *Recurrent Models of Visual Attention* was [published](https://arxiv.org/abs/1406.6247). 

> In
this paper we ... develop a novel framework for attention-based
task-driven visual processing with neural networks. Our model considers attention-based processing of a visual scene as a control problem ...
>
> The model is a recurrent neural network (RNN) which processes inputs sequentially, attending to
different locations within the images (or video frames) one at a time, and incrementally combines
information from these fixations to build up a dynamic internal representation of the scene or environment. Instead of processing an entire image or even bounding box at once, at each step, the model
selects the next location to attend to based on past information and the demands of the task.

## Cho et al. (2014) - Performance Deteriorates With Sequence Length

In September, 2014, Cho et. al. [published](https://arxiv.org/abs/1409.1259) *On the Properties of Neural Machine Translation: Encoder-Decoder Approaches* 

Referencing this paper in another paper named *Neural Machine Translation by Jointly Learning to Align and Translate*, [published](https://arxiv.org/abs/1409.0473) in the same month, Cho, Bengio, and Bahdanau notes

> . Cho et al. (2014b) showed that indeed the performance of
a basic encoder–decoder deteriorates rapidly as the length of an input sentence increase


## Cho, Bengio, and Bahdanau (2014) - Decoder Attention For RNNs
In September 2014, a paper titled *"Neural machine translation by jointly learning to align and translate"* was [published](https://arxiv.org/abs/1409.0473).

Bahdanau et al. acknowlege the issues of the fixed length vector:

> The models proposed recently for neural machine translation often belong to a family of encoder-decoders and consists of an encoder that encodes a source sentence into a fixed-length vector from which a decoder generates a translation. 

They continue:

> The whole encoder–decoder system, which consists of the encoder and the decoder for a language pair, is jointly trained to maximize the probability of a correct translation given a source sentence. A potential issue with this encoder–decoder approach is that a neural network needs to be able to compress all the necessary information of a source sentence into a fixed-length vector. This may make it difficult for the neural network to cope with long sentences, especially those that are longer than the sentences in the training corpus.

Additionally they propose a novel enhancement to the encoder-decoder architecture which is built around an attention mechanism.

> In this paper, we conjecture that the use of a fixed-length vector is a bottleneck in improving the performance of this basic encoder-decoder architecture, and propose to extend this by allowing a model to automatically (soft-)search for parts of a source sentence that are relevant to predicting a target word, without having to form these parts as a hard segment explicitly.
>
> By letting the decoder have an attention mechanism, we relieve the encoder from the burden of having to encode all information in the source sentence into a fixedlength vector. With this new approach the information can be spread throughout ..., which can be selectively retrieved by the decoder accordingly.

They emphasize that this mechanism bypasses the bottleneck of a fixed length vector

> The most important distinguishing feature of this approach from the basic encoder–decoder is that it does not attempt to encode a whole input sentence into a single fixed-length vector. Instead, it encodes the input sentence into a sequence of (context) vectors and chooses a subset of these vectors adaptively (based on the alignment scores) while decoding the translation. This frees a neural translation model from having to squash all the information of a source sentence, regardless of its length, into a fixed-length vector. We show this allows a model to cope better with long sentences.

At the center of their proposed architecture are is an attention mechanism which relies on a context vector calculated as an expectation, or weighted sum of annotations.

For each token in the input sequence, the encoder (a bi-driectional RNN) will compute the annotations (a concatenation of the forward and backward states of the encoder's RNNs).

> In this way, the annotation $h_j$ contains the summaries of both the preceding words and the following words. Due to the tendency of RNNs to better represent recent inputs, the annotation $h_j$ will be focused on the words around $x_j$ . This sequence of annotations is used by the decoder and the alignment model later to compute the context vector.

The context vector is computed as the weighted sum of the annotations (represented by $\bigoplus$). In this case, the weights $\alpha$ are calculated as the softmax of the alingment value of the annotations. 

**Note**: Use of the softmax function continues forward into newer attention mechanisms.


This is then used as an input to the decoders hidden state $s_t$ to predict the target value $y_t.
$.
> The model then predicts a target word based on the context vectors associated with these source positions and all the previous generated target words.

<center><img src='./images/attention_architecture_example.png' style='width:25%'></center>

The model is trained by optimizing an alignment score; which quantifies the distance from the correct outputs given the inputs. 

The authors show how the attention scores allow the model to dynamically determine the soft-alignment of the input and output sequences. Specifically, they show several instances where the attention model is able to change the order of words from their exact literal translation to a gramatically correct translation.

<center><img src="./images/attention_swapping_word_order.png" style="width:50%"></center>

> Figure 3: Four sample alignments found by RNNsearch-50. The x-axis and y-axis of each plot
correspond to the words in the source sentence (English) and the generated translation (French),
respectively. Each pixel shows the weight $\alpha_{ij}$ of the annotation of the j-th source word for the i-th
target word (see Eq. (6)), in grayscale (0: black, 1: white). (a) an arbitrary sentence. (b–d) three
randomly selected samples among the sentences without any unknown words and of length between
10 and 20 words from the test set.
 


## Bahdanau et. al. (2015) - Global Hard And Soft Alignment & Additive Attention

In September 2014, Bahdanau et. al. [published](https://arxiv.org/abs/1409.0473) *Neural Machine Translation by Jointly Learning to Align and Translate*. 

The authors state

> In this paper, we show that the proposed approach of jointly learning to align and translate achieves
significantly improved translation performance over the basic encoder–decoder approach


They acknowledge that their work builds on the previous work of several researchers

> Here, we describe briefly the underlying framework, called RNN Encoder–Decoder, proposed by
Cho et al. (2014a) and Sutskever et al. (2014) upon which we build a novel architecture that learns
to align and translate simultaneously.

They give a definition of the basic encoder-decoder to clarify the term

> An encoder neural network reads and encodes a source sen-
tence into a fixed-length vector. A decoder then outputs a translation from the encoded vector. The
whole encoder–decoder system, which consists of the encoder and the decoder for a language pair,
is jointly trained to maximize the probability of a correct translation given a source sentence.

They then discuss the short comings of this basic encoder-decoder architecture

> A potential issue with this encoder–decoder approach is that a neural network needs to be able to
compress all the necessary information of a source sentence into a fixed-length vector. This may
make it difficult for the neural network to cope with long sentences, especially those that are longer
than the sentences in the training corpus.

Here, the fixed vector is referring to the hidden states in the last layer of the RNN. Traditionally, the activation function of the RNN only considered the previous hidden state from the previous iteration or time step. They were not considering multiple previous hidden states or all hidden states. Thus the RNN had a fixed memory and would "forget things" with longer sequences.

In their paper, Bahdanau et. al. suppose that a soft-search approach will bypass the fixed-length vector problem

> The models proposed recently for neural machine translation often belong to a family of encoder-decoders and consists of an encoder that encodes a source sentence into a fixed-length vector from which a decoder generates a translation. In this paper, we conjecture that the use of a fixed-length vector is a bottleneck in improving the performance of this basic encoder-decoder architecture, and propose to extend this by allowing a model to automatically (soft-)search for parts of a source sentence that are relevant to predicting a target word, without having to form these parts as a hard segment explicitly.

The soft-search approach ultimately relies on a soft-alignment. The authors highlight the differences between hard and soft alignment.

> Consider the source phrase [the man] which was translated into [l’ homme]. Any hard
alignment will map [the] to [l’] and [man] to [homme]. This is not helpful for translation, as one
must consider the word following [the] to determine whether it should be translated into [le], [la],
[les] or [l’]. Our soft-alignment solves this issue naturally by letting the model look at both [the] and
[man], and in this example, we see that the model was able to correctly translate [the] into [l’].

**Note**: The term alignment abstractly refers to the action or result of mapping sequences which literally translate into different order or alignment than the desired result. For more information see the [Extra Topics notebook](ExtraTopics.ipynb).

The solution they will ultimately propose side steps the fixed-length vector problem by introducing a mechanism which allows the decoder to selectively (i.e. softly) consider all of the hidden states in any number of hidden layers, regardless of how long the input sequence is.

This solution relies on an alignment model which scores how well the inputs match the outputs by attaching a weight to the coresponding hidden states of the RNN. 

This alignment model is proposed as a separate explicit model which is trained in tandem with the encoder-decoder.

> We parametrize the alignment model a as a feedforward neural network which is jointly trained with
all the other components of the proposed system. Note that unlike in traditional machine translation the alignment is not considered to be a latent variable. Instead, the alignment model directly computes a soft alignment, which allows the gradient of the cost function to be backpropagated through.
This gradient can be used to train the alignment model as well as the whole translation model jointly.

**Note**: Becuase we are selectively considering the prior hidden states, we are esentially talking about an attention mechanism. Thus the attention mechanism is helping to provide the alignment. This leads to the terms alignment and attention getting conflated; while the authors use the term soft-alignment and not soft attention, because alignment and attention are now part of the same mechanism, we have various citations attributing the term soft-attention for example.

This is a major departure from previous implimentations.

> The attention mechanism proposed by Bahdanau takes two important divergences from the seq2seq approach. First, there is a new context vector $\vec c_t$ 
 computed for every step of the translation sequence. Second, that context vector $\vec c_t$
 is not simply one of the hidden states from the input sequences, but is intead a weighted combination of the hidden states from the input sequence. Most importantly, that relative weighting is not static, but will change at every time step $t$
 so that relative importance of the input words can be dynamically adjusted depending on which word in the output sequence is being translated. Said differently, as the output translation proceeds from one word to the next, the context vector is re-weighting the input sequence hidden states in an attempt to highlight which input words are most important for a particular output word.
>
> [Hines - A Lit Review of Attentional Mechanisms](http://keeganhin.es/blog/attn.html)

Similarely in their conclusion they note:

> 

They are clear in their claims that their soft attention mechanism outperforms the prior hard attention implimentsions:

> <center><img src="images/monotonic_soft_attention.png" style="width:75%"></center>
>
> We can see from the alignments in Fig. 3 that the alignment of words between English and French
is largely monotonic. We see strong weights along the diagonal of each matrix. However, we also
observe a number of non-trivial, non-monotonic alignments. Adjectives and nouns are typically
ordered differently between French and English, and we see an example in Fig. 3 (a). From this
figure, we see that the model correctly translates a phrase [European Economic Area] into [zone
economique europ ´ een]. The RNNsearch was able to correctly align [zone] with [Area], jumping ´
over the two words ([European] and [Economic]), and then looked one word back at a time to
complete the whole phrase [zone economique europ ´ eenne]. ´
>
> The strength of the soft-alignment, opposed to a hard-alignment, is evident, for instance, from
Fig. 3 (d). Consider the source phrase [the man] which was translated into [l’ homme]. Any hard
alignment will map [the] to [l’] and [man] to [homme]. This is not helpful for translation, as one
must consider the word following [the] to determine whether it should be translated into [le], [la],
[les] or [l’]. Our soft-alignment solves this issue naturally by letting the model look at both [the] and
[man], and in this example, we see that the model was able to correctly translate [the] into [l’]. We
observe similar behaviors in all the presented cases in Fig. 3. An additional benefit of the soft alignment is that it naturally deals with source and target phrases of different lengths, without requiring a
counter-intuitive way of mapping some words to or from nowhere ([NULL]) (see, e.g., Chapters 4
and 5 of Koehn, 2010).

Note: The attention mechanism proposed in this paper will later be referred to as a type of global attention mechanism by Luong et. al. (2015). Luong notes that while more complicated than the global attention mechanism he proposes, in the field of NMT, 
Bahdanau was first to propose global attention.

>To the best of our knowledge, there has not been any other work exploring the use of attention-based architectures for NMT.

To clarify the meaning of "global" we have this [post](http://keeganhin.es/blog/attn.html):

> The model put forth by Bahdanau uses what they call global attention: every single word of the input sequence is in consideration for the attention vector.
>
> In the Bahdanau work, the attention vector is length-N vector which spans the entire input sequence. Thus, every single word of the input sequence receives at least some amount of weight and the attentional vector is somewhat smoothed over the input sequence. An alterative idea would be to select a subset of the words which receive some attentional weight whereas the rest will receive exactly zero weight. This "hard" attention indudes a sparsity in the attention vector and allows more precise attending. This sparse mechanism is no longer differentiable

Additionally, we see various posts ([example](http://keeganhin.es/blog/attn.html)) refer to this attention mechanism as additive attention. This is due to the mathematical formula which sums up the products of the attention weights and the coresponding annotations (concatenated hidden states of the encoder).

## Xu et. al. (2015) - Applying Soft/Hard Attention To Image Processsing

In Feb 2015, Xu et. al. [published](https://arxiv.org/abs/1502.03044) *Show, Attend and Tell: Neural Image Caption Generation with Visual Attention*.

> Encouraged by recent advances in caption generation and inspired by recent success in employing attention
in machine translation (Bahdanau et al., 2014) and object
recognition (Ba et al., 2014; Mnih et al., 2014), we investigate models that can attend to salient part of an image while
generating its caption.

This paper introduces an attention based model that automatically learns to summarize the content depicted in images.

In the paper they describe two attention mechanisms which they apply to caption generation.

> We introduce two attention-based image caption generators under a common framework (Sec. 3.1): 1) a “soft” deterministic attention mechanism trainable by standard back-propagation methods and 2) a “hard” stochastic attention mechanism trainable by maximizing an approximate variational lower bound or equivalently by REINFORCE (Williams, 1992).

## Luong, et al. (2015) - Global vs. Local Attention

In August 2015, Luong et. al. [published](https://arxiv.org/abs/1508.04025) *Effective Approaches to Attention-based Neural Machine Translation*.

In this article they define two opposing architecures for attention mechanisms which are named local and global attention. Their abstract reads:

> An attentional mechanism has lately been
used to improve neural machine translation (NMT) by selectively focusing on
parts of the source sentence during translation. However, there has been little
work exploring useful architectures for
attention-based NMT. This paper examines two simple and effective classes of attentional mechanism: a global approach
which always attends to all source words
and a local one that only looks at a subset
of source words at a time. We demonstrate
the effectiveness of both approaches on the
WMT translation tasks between English
and German in both directions

I have seen [claims](https://lilianweng.github.io/posts/2018-06-24-attention) that they "proposed" the concepts and architecture. For example this quote:

> Luong, et al., 2015 proposed the “global” and “local” attention. The global attention is similar to the soft attention, while the local one is an interesting blend between hard and soft, an improvement over the hard attention to make it differentiable: the model first predicts a single aligned position for the current target word and a window centered around the source position is then used to compute a context vector.

I believe this is accurate; while the existing implimentations may be able to classify according to the new paradigms, the authors explain that they would need simplificiations if that was the case:

> In this work, we design, with simplicity and effectiveness in mind, two novel types of attention based models: a global approach in which all
source words are attended and a local one whereby
only a subset of source words are considered at a
time. The former approach resembles the model
of (Bahdanau et al., 2015) but is simpler architecturally. The latter can be viewed as an interesting
blend between the hard and soft attention models
proposed in (Xu et al., 2015): it is computationally less expensive than the global model or the
soft attention; at the same time, unlike the hard attention, the local attention is differentiable almost
everywhere, making it easier to implement and
train.


As highlighed in this [post](http://keeganhin.es/blog/attn.htm), the authors consider several similarity functions (denoted as $score()$) used by the softmax attention score.

Given a target hidden state $h_t$, and all source states $\bar h_s$ , 

* dot-product: $h_t^T \bar h_s$
* general: $h_t^T W_a\bar h_s$
* concat: $v_a^T \tanh(W_a[h_t;\bar h_s])$

The simplest of these is obviously the dot-product similarity which contains no additional free parameters.

## Britz et. al (1017) - Global Soft Attention Catches On

In March 2017, Britz [published](https://arxiv.org/abs/1703.03906) *Massive Exploration of Neural Machine Translation Architectures*. 

This paper is important because it analyzes variants of the existing NMT architectures which are based on the architecturs which utilize the attention patterns of Bahdanau and Luong.

> The most popular approaches
to NMT are based on an encoder-decoder architecture consisting of two recurrent neural networks
(RNNs) and an attention mechanism that aligns
target with source tokens (Bahdanau et al., 2015;
Luong et al., 2015a)
>
> ...
>
> Our models are based on an encoder-decoder architecture with attention mechanism (Bahdanau
et al., 2015; Luong et al., 2015a)

## Vaswani et al (2017) - Transformers and Attention

In June 2017, the transformer architecture was [published](https://arxiv.org/abs/1706.03762) by Vaswani et al. This new architecture incorporated and expanded upon the prior advancements with respect to attention. This model, based solely on attention mechanisms, dispenses with recurrence and convolutions entirely.